Core/Src/mcp3008.c

```diff
@@ -0,0 +1,43 @@
+/*
+ * MCP3008.c
+ *
+ */
+
+#include "../Inc/mcp3008.h"
+
+/*
+ * Set the MISO, MOSI, SCK and CS
+ * SPI settings:
+ * CPHA = 1 Edge
+ * Prescaler = 8
+ * First bit = MBS first
+ * CPOL = Low
+ */
+void MCP3008_Init(MCP3008_SPI* spi, SPI_HandleTypeDef* hspi, GPIO_TypeDef* CS_PORT, uint16_t CS_PIN){
+  spi->hspi = hspi;
+  spi->CS_PORT = CS_PORT;
+  spi->CS_PIN = CS_PIN;
+}
+
+// Read the channels from 0 to 7
+uint16_t MCP3008_Read_Channel(MCP3008_SPI* spi, uint8_t channel){
+
+	// Declare data that we will send
+	uint8_t pTxData[3] = {0};
+	pTxData[0] = ((0x01 << 7)|		  // start bit
+		     (1 << 6)|			// SGL
+		     ((channel & 0x07) << 3)); 	// channel number
+	pTxData[1] = 0x00;
+	pTxData[2] = 0x00;
+
+	// Data that we will get
+	uint8_t pRxData[3] = {0};
+
+	// CS low, Send and receive, CS high
+	HAL_GPIO_WritePin(spi->CS_PORT, spi->CS_PIN, GPIO_PIN_RESET);
+	HAL_SPI_TransmitReceive(spi->hspi, pTxData, pRxData, 3, 10);
+	HAL_GPIO_WritePin(spi->CS_PORT, spi->CS_PIN, GPIO_PIN_SET);
+
+	// Compute the ADC
+return 0x3FF & ((pRxData[0] & 0x01) << 9 | (pRxData[1] & 0xFF) << 1 | (pRxData[2] & 0x80) >> 7);

```

Comment:
I like the `0x3FF` mask. That ensures the output is in the valid range. 
The datasheet says the first received byte has no useful information, so you probably don't want to use it as well. The rest bit manipulation is wrong, I see you might have some misunderstanding with the bit operations here. For example, `(pRxData[2] & 0x80) >> 7` is not shifting the middle chunk a three-byte data by 7, but it is shifting a uint8_t type data right by 7, leaving only one bit from `pRxData[2]`, and the 7 bits before it are filled with zeros. 
I am also a bit concerned about the data type here. Though you specified uint16_t as the return type, everything to the right of the return is not specified with their type very well.

Reply:
I think it should be okay now